core/modules/taxonomy/taxonomy.module

```diff
@@ -1852,6 +1853,44 @@ function taxonomy_field_widget_form(&$form, &$form_state, $field, $instance, $la
   return $element;
 }
 
+/**
+ * Implements hook_field_widget_options_buttons_form_alter().
+ */
+function taxonomy_field_widget_options_buttons_form_alter(&$element, &$form_state, $context) {
+  if ($context['field']['type'] == 'taxonomy_term_reference') {
+    if (!empty($context['field']['settings']['allowed_values'])) {
+      $trees = $context['field']['settings']['allowed_values'];
+    }
```

Comment:
I don't understand why you are loading the trees from `['settings']['allowed_values']`. Is there any scenario when the options won't be available in `$element['#options']`? 

In any case, I think loading directly from `['settings']['allowed_values']` is not a good idea either, because a contrib module like `taxonomy_access` or `i18n_taxonomy` may have defined a callback within `['settings']['options_list_callback']` that is supposed to override `['settings']['allowed_values']`.  I believe `['settings']['allowed_values']` will still not be empty, even when there is a callback in `options_list_callback`. Which means that the code above will be ignoring that special callback.

Reply:
> Is there any scenario when the options won't be available in `$element['#options']`?

Not that I know of, but we need more than just the tids (which are in #options); we need the vocabularies themselves (which are in #allowed_values) in order to compute the term depth. (I originally took that code from `taxonomy_allowed_values()`, which was this:

```
function taxonomy_allowed_values($field) {
  $options = array();
  foreach ($field['settings']['allowed_values'] as $tree) {
    if ($vocabulary = taxonomy_vocabulary_load($tree['vocabulary'])) {
      if ($terms = taxonomy_get_tree($vocabulary->machine_name, $tree['parent'])) {
        foreach ($terms as $term) {
          $options[$term->tid] = str_repeat('· ', $term->depth) . $term->name;
        }
      }
    }
  }
  return $options;
}
```

I'd assumed (incorrectly, it appears) that if that callback was being overidden, that the `empty()` check above would detect whether I can rely on `allowed_values`, and save the extra db call(s) to get the term vocabularies.

At this point, though, it might be simpler to just use the current `else` clause for all cases to get the `$trees`. It's only one extra db call. On the other hand, for the vast majority of use cases, people won't be overriding the callback, so it would be save to use `allowed_values` to get the $trees.

What do you think?

Reply:
Meanwhile, I've updated the check for the override to check for the callback, not `allowed_values`.